### PR TITLE
[NAVSDK-757] optimize MapboxRouteLineApi#updateWithRouteProgress

### PR DESCRIPTION
**Pull request by LukasPaczos**
### Description

Alternative to https://github.com/mapbox/mapbox-navigation-android/pull/6646. Uses the current route segment index that we're on (from Nav Native) to calculate the vanishing offset.

Creating as a draft first as we'll need to update tests as well.

### CODE REVIEW

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/model/RoutePoints.kt [hunk 1]
```diff
@@ -10,5 +10,4 @@ import com.mapbox.geojson.Point
  */
 internal data class RoutePoints constructor(
     val stepPoints: List<List<List<Point>>>,
-    val flatList: List<Point>,
```
Comment (dzinad) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/model/RoutePoints.kt:13]: 1) we can remove this from kdoc;
2) do we need the data class at all in this case?

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineApi.kt [hunk 2]
```diff
@@ -1099,61 +1095,8 @@ class MapboxRouteLineApi(
...
-
-            /**
-             * When we know the number of remaining points and the number of all points,
-             * calculate the index of the upcoming point.
-             */
-            val allPoints = completeRoutePoints.flatList.size
-            routeLineOptions.vanishingRouteLine?.primaryRouteRemainingDistancesIndex =
-                allPoints - allRemainingPoints - 1
-        } ?: run { routeLineOptions.vanishingRouteLine?.primaryRouteRemainingDistancesIndex = null }
+        routeLineOptions.vanishingRouteLine?.upcomingRouteGeometrySegmentIndex =
```
Comment (dzinad) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineApi.kt:1098]: This looks... Much simpler.

File: libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/route/line/MapboxRouteLineUtils.kt [hunk 3]
```diff
@@ -723,7 +722,7 @@ internal object MapboxRouteLineUtils {
     ) -> RouteLineGranularDistances? =
         { route: NavigationRoute ->
             val points = routePointsProvider(route)
-            if (points?.flatList?.isNotEmpty() == true) {
+            if (points?.stepPoints?.isNotEmpty() == true) {
```
Comment (dzinad) [inline libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/route/line/MapboxRouteLineUtils.kt:725]: It's not exactly the same: stepPoints could be sth like: [[], []] which means they are not empty while flatLost would be empty.

Comment (LukasPaczos) [reply libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/route/line/MapboxRouteLineUtils.kt:725]: Yea, maybe it's better to refactor the receiving function to accept nullable and empty collections.

File: libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineApiRoboTest.kt [hunk 4]
```diff
@@ -428,13 +506,13 @@ class MapboxRouteLineApiRoboTest {
                 .vanishingRouteLineUpdateInterval(0)
                 .build()
             val api = MapboxRouteLineApi(options)
-            val expectedRestrictedExpression = "[literal, [0.0, 0.05416168943228483]]"
-            val route = loadNavigationRoute("route-with-restrictions.json")
+            val expectedRestrictedExpression = "[literal, [0.0, 0.04111412293617067]]"
```
Comment (LukasPaczos) [inline libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineApiRoboTest.kt:509]: This is the test case I was referring to in the comment above.

**Review by LukasPaczos — COMMENTED** [review]
While fixing up tests I think I found a solution for a problem that has been popping up occasionally but we couldn't narrow it down.

The problem was surfaced by this test after my changes - https://github.com/mapbox/mapbox-navigation-android/pull/6648/files#diff-f1c12e5630865905bb07e96bfeb22dc9be4929a6f82ae117da458ff5e453facdR509. The code was selecting a geometry index of 8 instead of 7 as the "upcoming" index which led to bigger offset than expected. Here are the changes that we'd have to apply on the `main` branch to fix this issue:
```diff
diff --git a/libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineApi.kt b/libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineApi.kt
index c3f4828207..d07a382bbd 100644
--- a/libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineApi.kt
+++ b/libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineApi.kt
@@ -1132,7 +1132,7 @@ class MapboxRouteLineApi(
                 val currentLegSteps = completeRoutePoints.stepPoints[currentLegProgress.legIndex]
                 allRemainingPoints += if (currentStepProgress.stepIndex < currentLegSteps.size) {
                     currentLegSteps.slice(
-                        currentStepProgress.stepIndex + 1 until currentLegSteps.size - 1
+                        currentStepProgress.stepIndex + 1 until currentLegSteps.size
                     ).flatten().size
                 } else {
                     0
@@ -1152,7 +1152,7 @@ class MapboxRouteLineApi(
              */
             val allPoints = completeRoutePoints.flatList.size
             routeLineOptions.vanishingRouteLine?.primaryRouteRemainingDistancesIndex =
-                allPoints - allRemainingPoints - 1
+                allPoints - allRemainingPoints
         } ?: run { routeLineOptions.vanishingRouteLine?.primaryRouteRemainingDistancesIndex = null }

         lastIndexUpdateTimeNano = System.nanoTime()
```
Essentially, we were not counting the last step in the "remaining points on the route" which led to the remaining points count being smaller than expected and this inflated the "upcoming" point index. This PR resolves the problem. I also found a test case that seems to confirm it:

_before_

https://user-images.githubusercontent.com/16925074/204020907-1bc633aa-6c97-4a55-95e1-80977f49bf64.mp4

_after_

https://user-images.githubusercontent.com/16925074/204020941-16204de8-0bbc-4544-bd02-b7182caf855f.mp4

**Review by cafesilencio — APPROVED** [review]

**Pull request merged**